### ISSUE TITLE
Representing a tuple in Expr

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#1213]
IM4 in manual SSP appears like so:

![image](https://user-images.githubusercontent.com/18634568/56385596-a3842280-61ed-11e9-9612-1298a32ccc49.png)

Note that the left-hand side of the equation is a tuple, which we can't currently represent in Drasil as an `Expr`. 

The underlying problem here may be that a single IM has more than one output, which I know @oluowoj encountered during her updates for GamePhys as well. I'm not sure what the resolution was for her case, if there was one.

**Comment by JacquesCarette** [comment]
Functions really ought to be able to produce a proper data-structure as output. That should be fairly easy to do.

What is a fair bit harder is to allow pattern-matching notation on the lhs of a definition to decompose that structure and assign each variable the right 'piece'. In a properly typed language, the technology to do this is understood. In Drasil, not so much.

For now, the best hack is to use 3 definitions: the function returns a 2-tuple, which is named, and then you use an explicit projection to extract the components. But this feature request really ought to be implemented too.

**Comment by bmaclach** [comment]
I want to make sure I'm understanding your suggested short-term solution correctly. Would the "explicit projections" be in the form of two new data definitions that relate the 2-tuple to its components?

**Comment by JacquesCarette** [comment]
Yes.  New DD (that should be quite generic, so maybe outside SSP) corresponding to `fst` and `snd` in Haskell.

**Comment by bmaclach** [comment]
Thanks @JacquesCarette . I wasn't thinking of generic DDs... I'm confused about how that would work. First, I don't see any precedent for defining DDs outside of a specific example in Drasil. More importantly, if I were to define some generic DDs for `fst` and `snd`, how would I then tell Drasil that these DDs apply to the 2-tuple I'm using in SSP?

What makes more sense to me, and maybe this is what you meant in the first place, would be to define some new symbols, outside SSP, for `fst` and `snd`, and then define DDs, in SSP, that use these new symbols to relate the 2-tuple to its components. The DD in SSP would be built from something like this:
```
critCoordsDDEq :: Expr
critCoordsDDEq = apply second [returnedTuple]
```
where `second` is the generic symbol defined somewhere outside SSP, and `returnedTuple` is the name of the 2-tuple. Then my only question would be where best to define `first` and `second`. But I might also be completely on the wrong track.

**Comment by JacquesCarette** [comment]
You are right that there is no precedent. It's just that none of the DDs were sufficiently general - although some of them might be pure 'physics' that ought to!

So you'd define new DDs in the hierarchy `Data.Drasil`, and then indeed 'import' them into SSP.

One could go with defining symbols outside SSP and then have the DDs just in SSP, that would work. But it feels like there isn't as much sharing of generic concepts happening then.

The difference, to me, is between
```haskell
fst :: (a,b) -> a
fst (x , _) = x
```
and 
```haskell
sspFst :: (Foo, Bar) -> Foo
sspFst (x , _) = x
```
We really ought to be able to define the first one generically and then re-use it directly, as we do in Haskell itself.

**Comment by bmaclach** [comment]
Okay, I understand the motivation for the generic DDs, I just have some questions about how this will look in practice.

First: After I import the generic DD to SSP, do I need to specialize it somehow? Or I just directly include the generic DD in the list of SSP's DDs?

Second: I'll need some generic symbols to define the generic DDs. A symbol to represent any generic 2-tuple, and symbols for the 2 elements of the tuple. Should these symbols just be (x,y), x, and y?

**Comment by JacquesCarette** [comment]
1. I would think just include the generic DD
2. I was thinking `pair`, and then `fst` and `snd` for the projection functions. But I might be answering the wrong question!

Maybe the problem here is that what is generic is not **data** but rather a pair of functions, `fst` and `snd`, and those are what need to be defined (and used).  But I see functions as data, so I what thinking you'd just define `fst` and `snd` as definitions.  Maybe we do need "function definitions" as separate from DDs.  [Sorry I didn't catch that earlier]

**Comment by bmaclach** [comment]
I agree that they should be separated from DDs - I've just remembered that DDs in Drasil are associated with equations where the LHS is just the data being defined. So if we tried to make a generic DD for `fst`, we would need an equation of the form `fst = something`  which really doesn't make sense given that, as you say, `fst` is a function.

So then my question is whether there are data-types currently in Drasil that we could use for these "function definitions". Currently data in drasil-data seems to be split into 3 categories: Units, Concepts, and Quantities. Certainly `fst` and `snd` are neither Units or Quantities, but could they be considered as Concepts? Could we define them as `ConceptChunk`s? Or do we need to design a new data-type to represent them? And if we do need a new data-type, does it make sense to actually move forward with implementing that considering this solution was originally intended as a temporary "hack"?
